flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStIncrementalSnapshotStrategy.java

```diff
@@ -0,0 +1,435 @@
+            transferBytes +=
+                    miscFilesTransferResult.stream()
+                            .mapToLong(HandleAndLocalPath::getStateSize)
+                            .sum();
+
+            HandleAndLocalPath manifestFileTransferResult =
+                    stateTransfer.transferFileToCheckpointFs(
+                            classifiedFiles.f3,
+                            snapshotResources.manifestFileSize,
+                            checkpointStreamFactory,
+                            stateScope,
+                            snapshotCloseableRegistry,
+                            tmpResourcesRegistry);
+            metaHandles.add(manifestFileTransferResult);
+            transferBytes += manifestFileTransferResult.getStateSize();
+
+            recordReusableHandles(sstHandles);
+
+            return transferBytes;
+        }
+
+        private void recordReusableHandles(List<HandleAndLocalPath> sstHandles) {
+            synchronized (uploadedSstFiles) {
+                switch (sharingFilesStrategy) {
+                    case FORWARD_BACKWARD:
+                    case FORWARD:
+                        uploadedSstFiles.put(
+                                checkpointId, Collections.unmodifiableList(sstHandles));
+                        break;
+                    case NO_SHARING:
+                        break;
+                    default:
+                        // This is just a safety precaution. It is checked before creating the
+                        // ForStIncrementalSnapshotOperation
+                        throw new IllegalArgumentException(
+                                "Unsupported sharing files strategy: " + sharingFilesStrategy);
+                }
+            }
+        }
+
+        private Tuple4<List<HandleAndLocalPath>, List<Path>, List<Path>, Path> classifyFiles() {
+            int totalFileNum = snapshotResources.liveFiles.size();
+
+            List<HandleAndLocalPath> transferredSstHandles = new ArrayList<>(totalFileNum);
+
+            List<Path> toTransferSstFiles = new ArrayList<>(totalFileNum);
+            List<Path> toTransferMiscFiles = new ArrayList<>(totalFileNum);
+            Path toTransferManifestFile = null;
+
+            for (Path filePath : snapshotResources.liveFiles) {
+                final String fileName = filePath.getName();
+
+                if (fileName.startsWith(MANIFEST_FILE_PREFIX)) {
+                    Preconditions.checkState(
+                            toTransferManifestFile == null,
+                            String.format("Backend:%s should only one manifest file.", backendUID));
+                    toTransferManifestFile = filePath;
+                } else if (fileName.endsWith(SST_FILE_SUFFIX)) {
+                    Optional<StreamStateHandle> uploaded =
+                            snapshotResources.previousSnapshot.getUploaded(fileName);
+                    if (uploaded.isPresent()) {
```

Comment:
Maybe you should check `checkpointStreamFactory.couldReuseStateHandle(uploaded.get())` as well. Please see `RocksIncrementalSnapshotStrategy#createUploadFilePaths`.

Reply:
Thanks for remainder !